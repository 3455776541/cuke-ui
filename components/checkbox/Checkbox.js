import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import cls from "classnames";

export default class Checkbox extends PureComponent {
  static defaultProps = {
    prefixCls: "cuke-checkbox",
    defaultChecked: false,
    indeterminate: false,
    checked: false,
    onChange: () => {}
  };
  static propTypes = {
    prefixCls: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    disabled: PropTypes.bool,
    checked: PropTypes.bool,
    defaultChecked: PropTypes.bool,
    indeterminate: PropTypes.bool
  };
  state = {
    checked: this.props.checked || this.props.defaultChecked
  };
  _onChange = e => {
    this.setState(({ checked }) => {
      return {
        checked: !checked
      };
    });
    this.props.onChange(e);
  };

  render() {
    const {
      className,
      prefixCls,
      disabled,
      children,
      value,
      indeterminate,
      ...attr
    } = this.props;
    const { checked } = this.state;
    return (
      <label className={cls(`${prefixCls}-wrapper`)} {...attr}>
        <span
          className={cls(prefixCls, className, {
            [`${prefixCls}-checked`]: checked,
            [`${prefixCls}-disabled`]: disabled,
            [`${prefixCls}-indeterminate`]: checked && indeterminate
          })}
        >
          <input
            type="checkbox"
            value={value}
            checked={checked}
            className={cls(`${prefixCls}-input`)}
            onChange={this._onChange}
            disabled={disabled}
          />
          <span className={cls(`${prefixCls}-inner`)} />
        </span>
        <span>{children}</span>
      </label>
    );
  }
}
