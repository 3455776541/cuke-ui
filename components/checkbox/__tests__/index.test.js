import assert from "power-assert";
import React from "react";
import { render, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import Checkbox from "../index";

describe("<Checkbox/>", () => {
  it("should render Checkbox", () => {
    const wrapper = render(
      <div>
        <Checkbox>黄瓜 ui</Checkbox>
        <Checkbox disabled>黄瓜 ui</Checkbox>
        <Checkbox checked disabled />
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("should render CheckboxGroup", () => {
    const wrapper = render(
      <Checkbox.Group>
        <Checkbox value="黄瓜">黄瓜</Checkbox>
        <Checkbox value="茄子">茄子</Checkbox>
        <Checkbox value="玉米">玉米</Checkbox>
        <Checkbox value="番茄" disabled>
          番茄
        </Checkbox>
      </Checkbox.Group>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("should render cuke-checkbox classnames", () => {
    const wrapper = render(
      <div>
        <Checkbox>黄瓜 ui</Checkbox>
        <Checkbox disabled>黄瓜 ui</Checkbox>
        <Checkbox checked disabled />
      </div>
    );
    assert(wrapper.find(".cuke-checkbox").length === 3);
    assert(wrapper.find(".cuke-checkbox-disabled").length === 2);
    assert(wrapper.find(".cuke-checkbox-checked").length === 1);
  });
  it("should can trigger change event", () => {
    const onChange = jest.fn();
    const wrapper = shallow(<Checkbox onChange={onChange}>黄瓜 ui</Checkbox>);
    wrapper.find("input").simulate("change");
    expect(onChange).toHaveBeenCalled();
  });
  it("should can not trigger change event", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Checkbox onChange={onChange} disabled>
        黄瓜 ui
      </Checkbox>
    );
    wrapper.find(".cuke-checkbox").simulate("click");
    expect(onChange).not.toHaveBeenCalled();
  });
  it("should can trigger change event for checkbox group", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Checkbox.Group onChange={onChange}>
        <Checkbox value="黄瓜">黄瓜</Checkbox>
        <Checkbox value="茄子">茄子</Checkbox>
        <Checkbox value="玉米">玉米</Checkbox>
        <Checkbox value="番茄" disabled>
          番茄
        </Checkbox>
      </Checkbox.Group>
    );
    wrapper
      .find(Checkbox)
      .at(0)
      .simulate("change", { target: {} });
    expect(onChange).toHaveBeenCalled();
  });
  it("should can not trigger change event for checkbox group", () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Checkbox.Group onChange={onChange} disabled>
        <Checkbox value="黄瓜">黄瓜</Checkbox>
        <Checkbox value="茄子">茄子</Checkbox>
        <Checkbox value="玉米">玉米</Checkbox>
        <Checkbox value="番茄">番茄</Checkbox>
      </Checkbox.Group>
    );
    setTimeout(() => {
      wrapper.find(".cuke-checkbox-input").simulate("change");
      expect(onChange).not.toHaveBeenCalled();
    });
  });
});
