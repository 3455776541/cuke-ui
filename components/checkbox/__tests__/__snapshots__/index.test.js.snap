// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox/> should render Checkbox 1`] = `
<div>
  <label
    class="cuke-checkbox-wrapper"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        class="cuke-checkbox-input"
        type="checkbox"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      黄瓜 ui
    </span>
  </label>
  <label
    class="cuke-checkbox-wrapper cuke-checkbox-disabled"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        class="cuke-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      黄瓜 ui
    </span>
  </label>
  <label
    checked=""
    class="cuke-checkbox-wrapper cuke-checkbox-checked cuke-checkbox-disabled"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        checked=""
        class="cuke-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span />
  </label>
</div>
`;

exports[`<Checkbox/> should render Checkbox.Button 1`] = `
<div>
  <label
    class="cuke-checkbox-wrapper cuke-checkbox-button-wrapper cuke-checkbox-default"
  >
    <span
      class="cuke-checkbox cuke-checkbox-button"
    >
      <input
        class="cuke-checkbox-input"
        type="checkbox"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      黄瓜 ui
    </span>
  </label>
  <label
    class="cuke-checkbox-wrapper cuke-checkbox-button-wrapper cuke-checkbox-disabled cuke-checkbox-default"
  >
    <span
      class="cuke-checkbox cuke-checkbox-button"
    >
      <input
        class="cuke-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      黄瓜 ui
    </span>
  </label>
  <label
    checked=""
    class="cuke-checkbox-wrapper cuke-checkbox-button-wrapper cuke-checkbox-checked cuke-checkbox-disabled cuke-checkbox-default"
  >
    <span
      class="cuke-checkbox cuke-checkbox-button"
    >
      <input
        checked=""
        class="cuke-checkbox-input"
        disabled=""
        type="checkbox"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span />
  </label>
</div>
`;

exports[`<Checkbox/> should render CheckboxGroup 1`] = `
<div
  class="cuke-checkbox-group"
>
  <label
    class="cuke-checkbox-wrapper"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        class="cuke-checkbox-input"
        type="checkbox"
        value="黄瓜"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      黄瓜
    </span>
  </label>
  <label
    class="cuke-checkbox-wrapper"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        class="cuke-checkbox-input"
        type="checkbox"
        value="茄子"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      茄子
    </span>
  </label>
  <label
    class="cuke-checkbox-wrapper"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        class="cuke-checkbox-input"
        type="checkbox"
        value="玉米"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      玉米
    </span>
  </label>
  <label
    class="cuke-checkbox-wrapper cuke-checkbox-disabled"
  >
    <span
      class="cuke-checkbox"
    >
      <input
        class="cuke-checkbox-input"
        disabled=""
        type="checkbox"
        value="番茄"
      />
      <span
        class="cuke-checkbox-inner"
      />
    </span>
    <span>
      番茄
    </span>
  </label>
</div>
`;
